File: models/ShogiModel.py
from dataclasses import dataclass, field
from enum import Enum
from typing import List, Dict



class PieceType(Enum):
    WANG = "Wang"
    CHANG = "Chang"
    SANG = "Sang"
    JA = "Ja"
    HOO = "Hoo"
    EMPTY = "Empty"


@dataclass
class Piece:
    pieceType: PieceType
    stayedTurns: int
    owner: int  # 1 or 2, 0이면 없음


@dataclass
class ShogiPlayer:
    userId: int
    userName: str
    playerId: int
    capturedPieces: List[Piece] = field(default_factory=list)


@dataclass
class BoardState:
    board: List[List[Piece]] = field(default_factory=list)
    currPlayerId: int = 1

    def initialize(self):
        width, height = 3, 4
        self.board = [[
            Piece(pieceType=PieceType.EMPTY, stayedTurns=0, owner=0)
            for _ in range(height)
        ] for _ in range(width)]

        # Player 1 (owner 1)
        self.board[0][0] = Piece(pieceType=PieceType.SANG, stayedTurns=0, owner=1)
        self.board[1][0] = Piece(pieceType=PieceType.WANG, stayedTurns=0, owner=1)
        self.board[2][0] = Piece(pieceType=PieceType.CHANG, stayedTurns=0, owner=1)
        self.board[1][1] = Piece(pieceType=PieceType.JA, stayedTurns=0, owner=1)

        # Player 2 (owner 2)
        self.board[2][3] = Piece(pieceType=PieceType.SANG, stayedTurns=0, owner=2)
        self.board[1][3] = Piece(pieceType=PieceType.WANG, stayedTurns=0, owner=2)
        self.board[0][3] = Piece(pieceType=PieceType.CHANG, stayedTurns=0, owner=2)
        self.board[1][2] = Piece(pieceType=PieceType.JA, stayedTurns=0, owner=2)

    # def getPlayerId(self) -> int:
    #     return self.currPlayerId

    
@dataclass
class SessionInfo:
    sessionId: int
    # userId1: ShogiPlayer
    # userId2: ShogiPlayer
    players: Dict[int, ShogiPlayer]
    board: BoardState = field(default_factory=BoardState)

    def __post_init__(self):
        self.board.initialize()

File: models/ShogiModel.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

[CreateAssetMenu(fileName="NewShogiPieceInfo", menuName="Shogi/PieceInfo")]
    public class ShogiPieceInfo : ScriptableObject
    {
        public PieceType pieceType;
        public Sprite pieceSprite;

        // 이동 규칙 예시: 상대적 이동 벡터들
        public List<Vector2Int> moveDirections;
    }

public enum PieceType
    { Wang, Chang, Sang, Ja, Hoo }
public class Piece
{
    public PieceType pieceType;
    public int stayedTurns;
    public int owner;
    public ShogiPieceInfo pieceInfo;
}
public class Cell
{
    public int x;
    public int y;
    public Piece piece;
}

public class ShogiBoard
{
    public int width;
    public int height;
    public Cell[,] cells;
    public void InitializeBoard(List<ShogiPieceInfo> pieceInfos)
    {
        width = 3;
        height = 4;
        cells = new Cell[width, height];
        for (int x = 0; x < width; x++)
        {
            for (int y = 0; y < height; y++)
            {
                cells[x, y] = new Cell();
                cells[x, y].x = x;
                cells[x, y].y = y;
                cells[x, y].piece = null;
            }
        }
        // Initialize cells
        {
            // Player 0
            cells[0, 0].piece = new Piece
            {
                pieceType = PieceType.Sang,
                stayedTurns = 0,
                owner = 0,
                pieceInfo = pieceInfos.Find(p => p.pieceType == PieceType.Sang)
            };
            cells[1, 0].piece = new Piece
            {
                pieceType = PieceType.Wang,
                stayedTurns = 0,
                owner = 0,
                pieceInfo = pieceInfos.Find(p => p.pieceType == PieceType.Wang)
            };
            cells[2, 0].piece = new Piece
            {
                pieceType = PieceType.Chang,
                stayedTurns = 0,
                owner = 0,
                pieceInfo = pieceInfos.Find(p => p.pieceType == PieceType.Chang)
            };
            cells[1, 1].piece = new Piece
            {
                pieceType = PieceType.Ja,
                stayedTurns = 0,
                owner = 0,
                pieceInfo = pieceInfos.Find(p => p.pieceType == PieceType.Ja)
            };
            // Player 1
            cells[2, 3].piece = new Piece
            {
                pieceType = PieceType.Sang,
                stayedTurns = 0,
                owner = 1,
                pieceInfo = pieceInfos.Find(p => p.pieceType == PieceType.Sang)
            };
            cells[1, 3].piece = new Piece
            {
                pieceType = PieceType.Wang,
                stayedTurns = 0,
                owner = 1,
                pieceInfo = pieceInfos.Find(p => p.pieceType == PieceType.Wang)
            };
            cells[0, 3].piece = new Piece
            {
                pieceType = PieceType.Chang,
                stayedTurns = 0,
                owner = 1,
                pieceInfo = pieceInfos.Find(p => p.pieceType == PieceType.Chang)
            };
            cells[1, 2].piece = new Piece
            {
                pieceType = PieceType.Ja,
                stayedTurns = 0,
                owner = 1,
                pieceInfo = pieceInfos.Find(p => p.pieceType == PieceType.Ja)
            };
        }
        
    }
}
public class ShogiPlayer
{
    public int userId;
    public string userName;
    public int playerId;
    public List<Piece> capturedPieces;
}

public class SessionInfo
{
    public int sessionId;
    public int userId1;
    public int userId2;
}
public class ShogiModel : MonoBehaviour
{
    public ShogiBoard board;
    public int turn;
    public int playerId;

    // Start is called before the first frame update
    void Start()
    {

    }

    // Update is called once per frame
    void Update()
    {

    }
}
